### PR TITLE
feat: add a useClientActions hook

### PR DESCRIPTION
**Pull request by moldy530**
# Pull Request Checklist

- [ ] Did you add new tests and confirm existing tests pass? (`yarn test`)
- [ ] Did you update relevant docs? (docs are found in the `site` folder, and guidelines for updating/adding docs can be found in the [contribution guide](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md))
- [ ] Do your commits follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [ ] Does your PR title also follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [ ] If you have a breaking change, is it [correctly reflected in your commit message](https://www.conventionalcommits.org/en/v1.0.0/#examples)? (e.g. `feat!: breaking change`)
- [ ] Did you run lint (`yarn lint:check`) and fix any issues? (`yarn lint:write`)
- [ ] Did you follow the [contribution guidelines](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md)?

---

## PR-Codex overview
This PR introduces a new `useClientActions` hook in the `aa-alchemy` package for executing client actions with a SmartAccountClient in React components.

### Detailed summary
- Added `useClientActions` hook for executing client actions
- Updated import statements and exports in related files
- Added documentation for `useClientActions` including usage and parameters

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by moldy530** [comment]
This is a WIP prototype not meant to be merged in yet... I wanted to prove something out: "Can we create a generic function that handles all of the plugin extension methods? Yes we can."

File: packages/alchemy/src/react/hooks/usePlugin.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  TFunctionName extends keyof TActions,
+  TActions extends { [x: string]: (...args: any[]) => unknown } = {
+    [x: string]: (...args: any[]) => unknown;
+  }
+> = {
+  functionName: TFunctionName;
+  args: Parameters<TActions[TFunctionName]>;
+};
+
+export function usePlugin<
```
Comment (moldy530) [inline packages/alchemy/src/react/hooks/usePlugin.ts:55]: TODO: jsdocs

Comment (moldy530) [reply packages/alchemy/src/react/hooks/usePlugin.ts:95]: actually... usePlugin isn't a good name for this hook. This hook is generalizable to using **any** client decorator. Not just the plugin hooks. I'll rename all of this

File: packages/alchemy/src/react/hooks/usePlugin.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  TTransport extends Transport = Transport,
+  TChain extends Chain | undefined = Chain | undefined,
+  TActions extends { [x: string]: (...args: any[]) => unknown } = {
+    [x: string]: (...args: any[]) => unknown;
+  }
+> = {
+  client?: UseSmartAccountClientResult<
+    TTransport,
+    TChain,
+    SupportedAccount<"MultiOwnerModularAccount">
```
Comment (moldy530) [inline packages/alchemy/src/react/hooks/usePlugin.ts:16]: if we make this generalizable then this should be the union type of supported accounts

File: packages/alchemy/src/react/hooks/usePlugin.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  isExecutingPluginAction: boolean;
+};
+
+export type ExecutableFunctionName<
+  TActions extends { [x: string]: (...args: any[]) => unknown } = {
+    [x: string]: (...args: any[]) => unknown;
+  }
+> = Exclude<
+  keyof TActions,
+  `encode${string}`
```
Comment (moldy530) [inline packages/alchemy/src/react/hooks/usePlugin.ts:44]: if we make this generalizable, then we should remove this `encode` filtering

**Comment by moldy530** [comment]
TODO: add docs for this

File: packages/alchemy/src/react/hooks/useClientActions.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  allArgs = ExecutableFunctionArgs<
+    TActions,
+    TFunctionName extends ExecutableFunctionName<TActions>
+      ? TFunctionName
+      : ExecutableFunctionName<TActions>
+  >
+> = {
+  functionName: TFunctionName;
+  args: allArgs;
+};
```
Comment (denniswon) [inline packages/alchemy/src/react/hooks/useClientActions.ts:65]: this is only meant to be used for modular accounts? then we should name this `useModularAccountClient`.
Also, we should also include checking if the plugin is installed first and in general, throw more informative state error back to the caller.

+ if this is only meant to be used for plugin actions, should name the hook accordingly.
Makes sense to group them and put in the subdir of hooks for modular account client actions.

Comment (moldy530) [reply packages/alchemy/src/react/hooks/useClientActions.ts:65]: No this is meant to be used with any client decorator, not just plugins. You can use the lightAccountActions decorator with this for example

File: packages/alchemy/src/react/hooks/useClientActions.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    ReturnType<TActions[keyof TActions]>,
+    Error,
+    ClientActionParameters<TActions, ExecutableFunctionName<TActions>>
+  >({
+    mutationFn: async <TFunctionName extends ExecutableFunctionName<TActions>>({
+      functionName,
+      args,
+    }: ClientActionParameters<TActions, TFunctionName>) => {
+      if (!client) {
+        // TODO: use the strongly typed error here
```
[the remark this answers is not in the document]
Comment (moldy530) [reply packages/alchemy/src/react/hooks/useClientActions.ts:112]: Throwing that client is not provided is not generic. This state has nothing to do with the accounts context whatsoever and everything to do with the nullable state of this hooks parameters -- `client`

**Review by denniswon — DISMISSED** [review]

**Review by rthomare — APPROVED** [review]

**Pull request merged**